content/blog/persisting-kubernetes-workloads-with-amazon-efscsi-volumes-using-pulumi-sdks/index.md

```diff
@@ -0,0 +1,397 @@
+
+```bash
+$ mkdir k8sapp && cd k8sapp
+
+$ pulumi new typescript
+
+$ npm install --save @pulumi/eks @pulumi/kubernetes
+
+```
+
+Let's now declare the storage class, pv, pvc and application pod to complete our k8s application set-up for this example. We will update `index.ts` file again and run `pulumi up`:
+
+```typescript
+import * as pulumi from "@pulumi/pulumi";
+import * as k8s from "@pulumi/kubernetes";
+
+const env = pulumi.getStack();
+const cluster = new pulumi.StackReference(`d-nishi/k8sinfra/${env}`);
+const kubeconfig = cluster.getOutput("kubeconfig");
+const efsFilesystemId = cluster.getOutput("efsFilesystemId");
+
+const k8sProvider = new k8s.Provider("cluster", {
+    kubeconfig: kubeconfig,
+ });
+
+//* STEP 5: Create a storage class, persistent volume and persistent volume claim
+
+export const storageclassEFS = new k8s.storage.v1.StorageClass("efs-sc", {
+    metadata: { name: "efs-sc" },
+    provisioner: "efs.csi.aws.com"
+  }, { provider: k8sProvider });
+  
+  const pvEFS = new k8s.core.v1.PersistentVolume("efs-pv", {
+  metadata: { name: "efs-pv" },
+  spec: { 
+    capacity: { storage: "5Gi" },
+    volumeMode: "Filesystem",
+    accessModes: [ "ReadWriteOnce" ],
+    persistentVolumeReclaimPolicy: "Recycle",
+    storageClassName: "efs-sc",
+    csi: { 
+      driver: "efs.csi.aws.com", 
+      volumeHandle: efsFilesystemId,
+    }
+  }
+  }, { provider: k8sProvider });
+  
+  const pvcEFS = new k8s.core.v1.PersistentVolumeClaim("efs-claim", {
+  metadata: { name: "efs-claim" },
+  spec: { 
+    accessModes: [ "ReadWriteOnce" ],
+    storageClassName: "efs-sc",
+    resources: { requests: { storage: "5Gi" } }
+  }
+  }, { provider: k8sProvider });
+
+//* STEP 6: Mount the endpoint to pod in EKS cluster
+
+export const newPod = new k8s.core.v1.Pod("efs-app", {
+  metadata: { name: "efs-app" },
+  spec: {
+    containers: [{ 
+      name: "app" , 
+      image: "centos", 
+      command: ["/bin/sh"],
+      args: ["-c", "while true; do echo $(date -u) >> /data/out.txt; sleep 5; done"],
+      volumeMounts: [{ 
+        name: "persistent-storage", 
+        mountPath: "/data",
+      }],
+    }],
+    volumes: [{
+      name: "persistent-storage",
+      persistentVolumeClaim: { claimName: "efs-claim" }
+    }],
+  }
+}, { provider: k8sProvider, dependsOn: pvcEFS });
+
+```
+
+Verify the pod is running and that data is being written into the EFS filesystem using:
+
+```bash
+$ kubectl exec -ti efs-app -- tail -f /data/out.txt
+Mon Jul 8 06:12:00 UTC 2019
+Mon Jul 8 06:12:05 UTC 2019
+Mon Jul 8 06:12:10 UTC 2019
+Mon Jul 8 06:12:15 UTC 2019
+
+```
+
+This brings us to the end of our solution with Pulumi and AWS EFS on Amazon EKS. For more examples, refer to Pulumi's open source repository [here](https://github.com/pulumi/examples). Refer to other Kubernetes blogs [here](https://blog.pulumi.com/author/nishi-davidson).
```

Comment:
The link "here" isn't especially useful for SEO purposes, also, rather than using direct-URLs prefer to use "shortcodes". e.g.

```
Refer to other [Kubernetes articles](https://www.pulumi.com/tags/kubernetes)
```

Reply:
done.